nixos/modules/services/hardware/kanata.nix

```diff
@@ -0,0 +1,173 @@
+
+  keyboard = { name, ... }: {
+    options = {
+      name = mkOption {
+        type = types.str;
+        default = name;
+        internal = true;
+        readOnly = true;
+      };
+      device = mkOption {
+        type = types.str;
+        example = "/dev/input/by-id/usb-0000_0000-event-kbd";
+        description = "Path to the keyboard device.";
+      };
+      enableCmd = mkEnableOption "kanata to execute commands";
+      config = mkOption {
+        type = types.lines;
+        example = ''
+          (defsrc
+            grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc
+            tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
+            caps a    s    d    f    g    h    j    k    l    ;    '    ret
+            lsft z    x    c    v    b    n    m    ,    .    /    rsft
+            lctl lmet lalt           spc            ralt rmet rctl)
+
+          (deflayer qwerty
+            grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc
+            tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
+            @cap a    s    d    f    g    h    j    k    l    ;    '    ret
+            lsft z    x    c    v    b    n    m    ,    .    /    rsft
+            lctl lmet lalt           spc            ralt rmet rctl)
+
+          (defalias
+            ;; tap within 100ms for capslk, hold more than 100ms for lctl
+            cap (tap-hold 100 100 caps lctl))
+        '';
+        description = ''
+          Configuration.
+          See <link xlink:href="https://github.com/jtroo/kanata"/> for more information.
+        '';
+      };
+    };
+  };
+
+  mkName = keyboard: "kanata-${keyboard.name}";
+
+  mkConfig = keyboard:
+    pkgs.writeText "${mkName keyboard}-config.kdb" ''
+      (defcfg
+        ${lib.strings.optionalString keyboard.enableCmd "danger-enable-cmd yes"}
+        linux-dev ${keyboard.device})
+
+      ${keyboard.config}
+    '';
+
+  mkService = keyboard: {
+    name = mkName keyboard;
+    value = {
+      description = "kanata for ${keyboard.device}";
+
+      # Because path units are used to activate service units, which
+      # will start the old stopped services during "nixos-rebuild
+      # switch", stopIfChanged here is a workaround to make sure new
+      # services are running after "nixos-rebuild switch".
+      stopIfChanged = false;
+
+      serviceConfig = hardeningOptions // {
+        DynamicUser = true;
+        SupplementaryGroups = with config.users.groups; [
+          input.name
+          uinput.name
+        ];
+      };
+      script = ''
+        ${cfg.package}/bin/kanata \
+          --cfg ${mkConfig keyboard}
+      '';
+    };
+  };
+
+  mkPath = keyboard: {
+    name = mkName keyboard;
+    value = {
+      description = "kanata trigger for ${keyboard.device}";
+      wantedBy = [ "multi-user.target" ];
+      pathConfig = {
+        PathExists = keyboard.device;
+      };
+    };
+  };
+
+  hardeningOptions = {
```

Comment:
it could be argued that since this is only used in one place it should be inlined where it's used.